src/cert/client_cert_store.cpp

```diff
@@ -43,24 +83,38 @@ void validate_certificate(const std::string& pem_cert)
     if (raw_cert == nullptr)
         throw std::runtime_error("invalid certificate data");
 }
+
+auto ensure_perms_for(const multipass::Path& cert_dir)
+{
+    QFile::setPermissions(cert_dir, QFile::ReadOwner | QFile::WriteOwner | QFile::ExeOwner);
+
+    return cert_dir;
+}
 } // namespace
 
-mp::ClientCertStore::ClientCertStore(const multipass::Path& cert_dir) : cert_dir{cert_dir}
+mp::ClientCertStore::ClientCertStore(const multipass::Path& cert_dir)
+    : cert_dir{ensure_perms_for(cert_dir)}, authenticated_client_certs{load_certs_from_file(cert_dir)}
 {
 }
 
 void mp::ClientCertStore::add_cert(const std::string& pem_cert)
 {
+    if (verify_cert(pem_cert))
+        return;
+
     validate_certificate(pem_cert);
     QDir dir{cert_dir};
     QFile file{dir.filePath(chain_name)};
-    auto opened = file.open(QIODevice::WriteOnly | QIODevice::Append);
-    if (!opened)
+    if (!MP_FILEOPS.open(file, QIODevice::WriteOnly | QIODevice::Append))
         throw std::runtime_error("failed to create file to store certificate");
 
-    size_t written = file.write(pem_cert.data(), pem_cert.size());
+    file.setPermissions(QFile::ReadOwner | QFile::WriteOwner);
+
+    size_t written = MP_FILEOPS.write(file, pem_cert.data(), pem_cert.size());
     if (written != pem_cert.size())
         throw std::runtime_error("failed to write certificate");
```

Comment:
Hmm, is that standard programming procedure?  We obviously don't do that much (if at all) in the other files we write to and even the approach you point out would be potentially be susceptible to file corruption.  Like if we wrote to a temp file, how would be know it's corrupt or not corrupt before writing that to the main file?

Reply:
Yeah, it is, because move is atomic. We do it in a few places (we should have abstracted it by now…) where I've noticed.

You're throwing here on errors - but you have already written to the file. Think disk empty, any kind of write error - the file is already corrupted. If instead you write to a temp file and only move on successful writes, the chance of corruption is much smaller.

Reply:
Ok, I see.  I think this operation should be abstracted out and used throughout the code though.

Reply:
After some discussion, it was determined that there are many places in the code where the corrupted file concern can occur and so a follow up PR should be done to introduce a new `utils` function that does the write to temp file & rename and fix all of the places at once.

Reply:
I found [`QSaveFile`](https://doc.qt.io/qt-5/qsavefile.html) which made this very straightforward.  I'll look into using this in other places where we do `QFile::write()` operations.